### DISCUSSION TITLE
Feature: bru_rerun but allowing new data as input (as "training" data)

### DISCUSSION THREAD

**Discussion by AlexandreBSimas** [inlabru-org/inlabru#245] [Ideas]
Hello,

I would like to know if it would be possible to either create a new `bru_rerun` that allows one to pass a new data set as training data, or to allow one to pass new data in `bru_rerun`. We would like this feature to be used in a cross_validation function. You can take a look of such a `cross_validation` function in https://github.com/davidbolin/rSPDE/blob/devel/R/inlabru_rspde.R. We tried to create one ourselves, but it is not stable under inlabru development (meaning that every now and then, when inlabru is updated, it breaks and we need to update it as well). 

Thank you very much!
Best,
Alexandre

**Comment by finnlindgren** [comment]
Hi, it would be helpful to know more precisely what changes affected your method; sometimes it's `INLA` changes that necessitate changes, and sometimes changes in what is put in what data structure; most recently, I can imagine your code being affected by the added support for special response objects, which changed how `inlabru` interacts with `inla.stack()`.  I assume you haven't had your function as part of the `rSPDE` cran release yet, as the reverse dependency checks I run on my machine haven't picked up any issues (you _do_ test these things in your package tests when INLA and inlabru are installed, right? ;) ).  I'm not sure if there's a way to add the `rSPDE` devel version to the checks run by `revdepcheck`.

But generally, this is a non-trivial problem that I think requires a refactoring of some of the `bru()` and `like()` code; One needs to handle both the `data` and `response_data` inputs, _and_ rerun parts of the code in the `like()`, such as the code that evaluates the response variable based on the data (which can now be `inla.mdata()` and `inla.surv()` objects), as well as the `E`, `Ntrials`, `scale`, and `weights` arguments, as those are evaluated in the context of the input data.  Then there are also data-modifying models, such as "cp", (where the whole "cross-validation" principle is different).

As long as there's a clear notion of "keep the latent model intact, but update the data inputs" (and potentially `domain`/`samplers`/`ips` as well, as that would be needed for the "cp" case) that should be doable with a code refactor that splits the `like()` code into "store info about input expressions" and "evaluate those expressions on some data". That way, the "update observation model based on new data" could be identical in the "initial run" and "cross validation" runs.

The most basic solution I think is to rerun the original `like()` calls with the new data, and replace the `lhoods` list with the new objects; the component mappers will have already been initialised based on the original data, so shouldn't be changed (this assumes the idea is to keep the latent model structure fixed; the new data will need to be compatible with the existing samplers, just as for prediction).  But a cleaner solution would be for the `bru_like` class to remember how to evaluate the various data-dependent objects, and have a `bru_like_update()` method or similar.

The second problem is how to initialise the "rerun" part with the correct linearisation states. Since the `inla` specific data structure changes when the data changes (size of predictors) changes, in this use case one would instead want to initialise on the `bru` side of things instead. I'm not sure which parts of the `iinla()` initialisation code assumes the model structure from a previous result is unchanged; would be easiest if the cross-validation runs only need initialising with the linearisation state, and not having to deal with initialising the hyperparameters.

Writing a generally applicable cross-validation method for `inlabru` models is hard, since there's an infinite set of models where the semantics are known only to the user, such as aggregated data models, where the link between input data and response data is determined solely by the general formula expression evaluation; what to include in `data` and `response_data` depends entirely on how the user has setup their data (and then there are the "cp" point process models, that have similar issues).  Unfortunately some of the code required to support those features needs to be involved for _all_ models, which is why implementing the above requires some understanding of those features even if one isn't trying to explicitly support such models.

PS. I will be on annual leave until mid-August, so inlabru development will be dormant in the coming weeks.

**Reply by finnlindgren** [reply]
An alternative approach might be one where one takes the existing `bru_like` data structure and just replaces parts of the response variable (the "leave-out" part) with `NA`. That would restrict the feature to "leave-out" cross-validation, since it would leave all the covariate inputs intact and just disable parts of the response data.  ("cp" type models would still need special handling; setting up the initial data to support proper leave-space-out point pattern cross validation)

This could be a feature of a `bru_like_update()` method as well; a bit easier to implement than the full data replacement approach. I think it would just involve updating the contents of `x$response_data[[x[["response"]]]]` in each `bru_like` object. These can be `inla.mdata()` and `inla.surv()` objects, but would normally be plain `numeric` vectors.

**Reply by AlexandreBSimas** [reply]
Hi Finn, 

Thank you very much for the replies!

Ahh, we have some tests with inla and inlabru, and also the vignettes. Every time we push, it builds everything with the most recent version of INLA and inlabru, and this is how we catch these errors when updates occur. 

Our goal is to have a cross validation that is general enough for our purposes, but you are right, definitely not the most general, and does not cover "cp" models, for instance. 

I really liked your idea in this reply, the "alternative approach" it is really very very good, and should be more robust with respect to future changes. Will implement it!

Once again, thank you very much!
Best,
Alexandre

**Reply by finnlindgren** [reply]
Ok! The "alternative approach" is a good first step, and quite robust I think. It should also be simple to make work with inla.mdata and inla.surv I think. If you make something in rSPDE we can transition the basic "set a certain part of the response to NA" method when I'm back to inlabru work (or you can start a feature branch/fork that can be later massaged&merged), so it can be supported longer term even if the internal inlabru storage changes again.  I like to try to plan interface aspects like that, so the code that needs to know about internal storage lives close to where that storage is defined, and external code shouldn't need to know...

**Reply by AlexandreBSimas** [reply]
Hi Finn,

It is a really good interface structure indeed.

With respect to the code, perfect! I will fork it and come up with a suggestion for an initial version of a `bru_like_update()` method, that can later be improved. I will also try to make it work with inla.mdata and inla.surv. 

I will let you know once I have it done. I will also not work on it now, as I have some other things to finish first, but I hope to have an initial version ready by mid august.

Best,
Alexandre
